course-outline.md

```diff
@@ -0,0 +1,103 @@
+5. Sync your local and origin remote with an upstream remote.
+    - This is much easier now with the new GitHub interface so will not take much time
+6. Use git reset to move a branch back in the commit history.
+    - This might be too much in a single lecture, but there is no natural break earlier
+    - Additional resources for all git chapters:
+        - https://pages.github.ubc.ca/mds-2021-22/DSCI_521_platforms-dsci_students/materials/lectures/1-lecture1-MDStools-shell-git-github.html#introduction-to-local-and-remote-version-control-software
+        - https://datasciencebook.ca/Getting-started-with-version-control.html
+        - https://merely-useful.tech/py-rse/git-cmdline.html
+
+## 6. Project organization and environments
+
+1. Organize projects and name files in a sound manner.
+    - [521 lec 7](https://pages.github.ubc.ca/mds-2021-22/DSCI_521_platforms-dsci_students/materials/lectures/7-ssh-filenames-project-organization.html#filenames-best-practices)
+2. Use virtual conda environments for Python.
+    - [521 lec 8](https://pages.github.ubc.ca/mds-2021-22/DSCI_521_platforms-dsci_students/materials/lectures/8-asking-effectively-and-virtual-environments.html#virtual-environments)
+
+## 7. JupyterLab and reproducible documents
+
+1. Be comfortable using JupyterLab
+    - [The first assignment in the programming course](https://github.ubc.ca/ubc-mds/prog-python-data-science-instructor/blob/master/source/assignment1/assignment1.ipynb) includes the Jupyter intro from lab 0 in 521.
+        - But I still think it is fine if we repeat some things and have an easy first assignment
+2. Create reproducible literate Python (e.g., reports and presentations) using Jupyter
+3. Write Markdown syntax within notebooks and markdown files
+4. There should be code and outputs in the notebook so that they get practice version controlling them.
+5. Advanced Jupyter
+    - Extensions?
+    - Jupytext?
+
+## 8. Creating books and web sites from Jupyter Notebooks
+
+1. Learning the basis of Jupyter-book
+2. Getting more practice version controlling notebooks
+3. There should be code and outputs in the notebook so that they get practice version controlling them.
+    - I think it would be really cool if students final project was to create a Jupyter Book, where they need to have a certain number of commits and make sure that it is published on GitHub.
+        - They have to practice essentially all they learned in this course to get it right!
+        - And it would be very rewarding for them to have made a book by the end of the program.
+
+
+## Optional extras if we need more content
+
+- Scaling up: read-eval-print-loop (REPL) processes versus non-interactive scripts
+- Write regular expressions for pattern matching. Maybe inside Python also.
+- Advanced bash with piping etc https://merely-useful.tech/py-rse/bash-tools.html
+- Setup and authenticate ssh keys
+- Makefiles
+- Writing script in VS code?
```

Comment:
@armanawn and I talked about this briefly today and agree that it would be a good idea to show some Python coding features of VSCode in the beginning, so that students understand that it is suitable for code in addition to markdown. For example, editing Python scripts and maybe showing how to divide code into cells with `%%`.

Including SSH with git could be a good best practice and we could include something simple like the lab assignment from 521 where they screenshot a successful authentication in their terminal. We don't want to spend too much time on this in the beginning, so potentially it could go later in the course or if there is one git chapter that too short it could fit there as well, since students would later be working with git in the following chapters (at least some of them).